air/src/execution_step/boxed_value/stream_map.rs

```diff
@@ -0,0 +1,188 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use std::rc::Rc;
+
+use super::ValueAggregate;
+use crate::execution_step::ExecutionResult;
+use crate::JValue;
+
+use super::stream::*;
+
+use air_interpreter_data::GenerationIdx;
+use air_trace_handler::merger::ValueSource;
+use air_trace_handler::TraceHandler;
+use serde::Serialize;
+use serde_json::json;
+
+#[derive(Debug, Default, Clone)]
+pub struct StreamMap {
+    stream: Stream,
+}
+
+impl StreamMap {
+    fn envelope(key: &(impl Into<JValue> + Serialize), value: Rc<JValue>) -> Rc<JValue> {
+        Rc::new(json!({ "key": key, "value": value }))
+    }
+
+    pub(crate) fn from_generations_count(previous_count: GenerationIdx, current_count: GenerationIdx) -> Self {
+        Self {
+            stream: Stream::from_generations_count(previous_count, current_count),
+        }
+    }
+
+    pub(crate) fn from_value(key: &(impl Into<JValue> + Serialize), value: ValueAggregate) -> Self {
+        let obj = StreamMap::envelope(key, value.result);
+        Self {
+            stream: Stream::from_value(ValueAggregate::new(obj, value.tetraplet, value.trace_pos)),
+        }
+    }
+
+    pub(crate) fn insert(
+        &mut self,
+        key: &(impl Into<JValue> + Serialize),
+        value: ValueAggregate,
+        generation: Generation,
+        source: ValueSource,
+    ) -> ExecutionResult<GenerationIdx> {
+        let obj = StreamMap::envelope(key, value.result);
+        self.stream.add_value(
+            ValueAggregate::new(obj, value.tetraplet, value.trace_pos),
+            generation,
+            source,
+        )
+    }
+
+    pub(crate) fn compactify(self, trace_ctx: &mut TraceHandler) -> ExecutionResult<GenerationIdx> {
+        self.stream.compactify(trace_ctx)
+    }
+
+    pub(crate) fn get_mut_stream_ref(&mut self) -> &mut Stream {
+        &mut self.stream
+    }
+
+    pub(crate) fn get_stream_ref(&mut self) -> &Stream {
+        &self.stream
+    }
+}
+
+use std::fmt;
+
+impl fmt::Display for StreamMap {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.stream.is_empty() {
+            return write!(f, "[]");
+        }
+
+        writeln!(f, "[")?;
+        for (id, generation) in self.stream.get_values_iter().enumerate() {
+            write!(f, " -- {id}: ")?;
+            for value in generation.iter() {
+                write!(f, "{value:?}, ")?;
+            }
+            writeln!(f)?;
+        }
+
+        write!(f, "]")
+    }
```

Comment:
This logic is already implemented in `Display` for streams, here you just need to print an underlying stream.

Reply:
Done